Add hxcoro.ds.Channel

Simn opened the pull request:
Adds the Channel implementation from https://github.com/Aidan63/haxe/issues/4#issuecomment-2836786224 because I always forget where to find it!

The only change I made is that `suspendedWrites` holds a `SuspendedWrite` instead of closures because I don't like storing closures if it can be helped. I also tried to add some documentation but I never really know what to say about such classes.

I also can't quite figure out the purpose of `writeQueue`. I guess this is an optimization?

Aidan63 commented:
writeQueue is an optimisation which allows the writer to not suspend if there is no pending read and the cost of some memory (storing it in the array). Eventually I imagined these queue sizes / behaviour would be definable through the constructor, but I just hardcoded it as part of the demo.

I think we should get rid of those mutexes for now, it was a failed bodge on my part. Most of our stuff isn't thread safe at the moment so they serve no real purpose. We'll also need a coroutine aware mutex implementation since mutex locks based on the calling thread whereas we'd want it to lock on the coroutine.

Simn commented:
Fair point, I've removed the mutex.

On my silly benchmark, I get slightly better results with `maxQueueSize = 1`. It's not very conclusive, but it certainly doesn't seem worse. That could depend on a number of factors though.

I also don't know what python's problem is.

Edit: There is some strange push/shift interaction here which has become visible because I removed the closure. I made it `return write.v` directly instead of pushing it onto the array and then having the code below shift it again. Although I'm now realizing that this is shift, not pop, so this is actually not correct either. Argh.

Aidan63 commented:
The write queue can help if the writer produces values faster than the reader can consume them. Instead of the writer immediately getting suspending the value is put on the queue and the writer is then immediately free to do other stuff (e.g. preparing the next value).

I remember seeing some odd behaviour with it as well on some targets. I didn't solve it at the time as I couldn't immediately see the problem so left it as I mainly threw that code together as a quick sample to show.

Simn commented:
Something strange is happening with hoisting values on python. I'm seeing this output:

```python
                elif (_hx_continuation1 == 19):
                    tmp1 = _hx_continuation._hx_hoisted10122.pop
                    _hx_continuation._hx_hoisted10123 = tmp1(0)
                    _hx_continuation.gotoLabel = 20
                elif (_hx_continuation1 == 20):
                    _hx_continuation._hx_hoisted10123 = None
                    _hx_continuation.gotoLabel = 21
                elif (_hx_continuation1 == 21):
                    _hx_continuation._hx_hoisted10123 = _hx_continuation._hx_hoisted10123
                    haxe_Log.trace("shift",_hx_AnonObject({'fileName': "hxcoro/ds/Channel.hx", 'lineNumber': 67, 'className': "hxcoro.ds.Channel", 'methodName': "read", 'customParams': [_hx_continuation._hx_hoisted10123]}))
                    _hx_continuation.gotoLabel = (22 if ((_hx_continuation._hx_hoisted10123 is None)) else 24)
```

State 20 somehow unsets our hoisted value, and then state 21 sets it to itself. I don't know why this would be python-specific though.

Edit: The debug output matches this:

```haxe
                        case 19: {
                                _hx_continuation._hx_hoisted10123 = python.Syntax.call(python.Syntax.field(_hx_continuation._hx_hoisted10122, "pop"), [0]);
                                _hx_continuation.gotoLabel = 20;
                        };
                        case 20: {
                                _hx_continuation._hx_hoisted10123 = cast null;
                                _hx_continuation.gotoLabel = 21;
                        };
                        case 21: {
                                _hx_continuation._hx_hoisted10123 = _hx_continuation._hx_hoisted10123;
                                haxe.Log.trace("shift", {fileName : "hxcoro/ds/Channel.hx", lineNumber : 67, className : "hxcoro.ds.Channel", methodName : "read", customParams : [_hx_continuation._hx_hoisted10123]});
                                if (_hx_continuation._hx_hoisted10123 == null) _hx_continuation.gotoLabel = 22 else _hx_continuation.gotoLabel = 24;
                        };
```

I think this must be related to python's `shift` in some way.

Simn commented:
Compiling this to C++ gives me a segfault quite often:

```haxe
function main() {
	for (size in [10000, 20000, 40000, 80000]) {
		var now = haxe.Timer.stamp();
		var scheduler = new EventLoopScheduler();
		final channel = new Channel();
		var task = CoroRun.with(scheduler).create(node -> {
			for (_ in 0...size) {
				node.async(_ -> {
					delay(Std.random(100));
					channel.write(2);
				});
			}
			var k = 0;
			for (_ in 0...size) {
				node.async(_ -> {
					delay(Std.random(100));
					k += channel.read();
				});
			}
			k;
		});
		task.start();
		scheduler.run();
		trace('size: $size, time: ${haxe.Timer.stamp() - now}');
	}
}
```

Occasionally it runs normally and prints what I expect it to print:

```
source/Main.hx:168: size: 10000, time: 0.2672615
source/Main.hx:168: size: 20000, time: 0.6295431
source/Main.hx:168: size: 40000, time: 1.3497426
source/Main.hx:168: size: 80000, time: 3.0613257
```

Simn commented:
Having profiled this a bit, I wasn't very impressed with the Array performance. I think `shift()` in particular is quite bad on several targets because it removes the first element and then mem-copies the array tail, which is not great. I came up with a `PagedDeque` structure that's some linked `Vector<T>` "pages", which are reused whenever possible to avoid allocations. I'm not sure if this is how real data structure programmers would approach this, but our targets sure appreciate the change:

js | before | after | before + fix | after + fix
--- | --- | --- | --- | ---
10000 | 0.420 | 0.384 | 0.227 | 0.227
20000 | 1.221 | 0.687 | 0.511 | 0.262
40000 | 2.235 | 1.343 | 1.267 | 0.438
80000 | 7.733 | 2.821 | 4.817 | 0.968

jvm | before | after | before + fix | after + fix
--- | --- | --- | --- | ---
10000 | 0.263 | 0.219 | 0.174 | 0.171
20000 | 0.467 | 0.430 | 0.194 | 0.134
40000 | 0.967 | 0.867 | 0.303 | 0.207
80000 | 2.143 | 1.710 | 0.659 | 0.277

hl | before | after | before + fix | after + fix
--- | --- | --- | --- | ---
10000 | 0.811 | 0.830 | 0.732 | 0.736
20000 | 1.610 | 1.665 | 1.492 | 1.385
40000 | 3.529 | 3.394 | 2.797 | 2.658
80000 | 7.399 | 6.848 | 5.855 | 5.271

cpp | before | after | before + fix | after + fix
--- | --- | --- | --- | ---
10000 | 0.293 | 0.259 | 0.182 | 0.172
20000 | 0.587 | 0.609 | 0.274 | 0.244
40000 | 1.380 | 1.264 | 0.567 | 0.458
80000 | 3.117 | 2.600 | 1.543 | 0.929

JS in particular is pretty crazy, this almost makes me think that something is wrong... CPP still segfaults like crazy, but it did that before my change, too.

Simn commented:
Added more numbers after improving event loop scheduling in https://github.com/Aidan63/haxe/commit/17e5ddf9955d94bb1f9875021791b3b96d6d39d8.

All targets still like this change. It feels like C++ segfaults even more now, I could barely get a run working. 

Edit: JProfiler before/after the scheduling change:

<img width="821" alt="jprofiler_KnDb4TWsyF" src="https://github.com/user-attachments/assets/99d6e9e2-fc36-4e3c-be49-9cacc0552b4d" />

<img width="821" alt="eL2elIZIw7" src="https://github.com/user-attachments/assets/898f8881-b681-4241-86b3-f978e79e8ae9" />

Slightly more balanced!

Aidan63 commented:
Is it that sample you posted a few comments up which is segfaulting? I just compiled and ran that about 20 times with the latest windows builds from this branches CI and all ran fine.

Simn commented:
I've updated my hxcpp and now I can't reproduce it anymore either. I must have been on some weird commit or something.
